### DISCUSSION TITLE
DropoutAdapter injection is impacting state_dict().keys and the ability to load/save checkpoints smoothly

### DISCUSSION THREAD

**Discussion by piercus** [finegrain-ai/refiners#237] [Ideas]
Hello refiners,

I'm experimenting with trainer and especially I'm facing a problem to load/save models weights 

The sequence of the trainer is :

1. `trainer.prepare_models` is loading the checkpoint on a non-injected model
2. `on_train_begin` is injecting the dropout_adapter
3. `on_checkpoint_save` is saving the checkpoint (using `model.state_dict()`)

The named of the Dropout-impacted layers are changed in step 2.

As a result, the model saved in `on_checkpoint_save` are not compatible with the loading in `trainer.prepare_models`, and i cannot smootly save/load the model.

## Toy example

The injection of the dropout adpater is changing the keys of weights in `state_dict()`

```py
from refiners.fluxion.layers.chain import Chain
from refiners.fluxion.layers.linear import Linear
from refiners.training_utils.dropout import DropoutAdapter

network = Chain(
    Linear(2, 3)
)

keys = network.state_dict().keys()
print(keys)

probability=0.5

for linear, parent in network.walk(Linear):
    DropoutAdapter(target=linear, probability=probability).inject(parent)

keys2 = network.state_dict().keys()
print(keys2)
```
is outputing
```
odict_keys(['Linear.weight', 'Linear.bias'])
odict_keys(['DropoutAdapter.Linear.weight', 'DropoutAdapter.Linear.bias'])
```

What i'm not clear is what is the target behavior 
A. should `.inject(parent)` change the name of the weights and we should fix the save/load sequence in the trainer ?
B. should `.inject(parent)` not change the name of the weights in `state_dict()` when the adapter is not injecting new weights ?

I can help on this if needed

**Comment by catwell** [comment]
We have an internal issue tracker, and the oldest issue I opened on it that is still open (before we even opened the source code of refiners) is: "Find a way to avoid changing the state dict when we add blocks without weights" :)

So yes this is something we have been discussing a lot internally. We don't really have a perfect solution for this now and there are always workarounds:

- eject the adapter before the snapshot and re-inject afterwards;
- make a [structural copy](https://github.com/finegrain-ai/refiners/blob/ab63a0158237b72fedc28ea1a4b0fcccdd2ca472/src/refiners/fluxion/layers/chain.py#L460) of the model, inject the adapter in one and use it to train, use the other one to snapshot;
- just post-process the state dict to remove DropoutAdapter from key names.


But if you have ideas it could be great!

Ideally I'd like the solution to also work when we e.g. insert a chain in a model for clarity / easier targeting, and to keep somewhat semantics keys (i.e. not just use for instance ordered keys named 0001 0002 etc, which works but had other issues).

**Reply by piercus** [reply]
> Ideally I'd like the solution to also work when we e.g. insert a chain in a model for clarity / easier targeting

Do you mean to say that, in a Chain, the key of a Linear layer should stay the same if I insert a Identity() layer before it ?

**Reply by catwell** [reply]
I think it is already the case. But what I meant is weights should still be compatible ir we turn e.g. this (pseudocode):

```py
Chain(
    Linear(),
    Conv(),
    Linear(),
    Conv(),
)
```

into this:

```py
Chain(
    Chain(
        Linear(),
        Conv(),
    ),
    Chain(
        Linear(),
        Conv(),
    ),
)
```

I know it is far from trivial to do this! It will almost certainly need a complete redesign of the keys scheme.

**Reply by piercus** [reply]
I'd like to suggest building a `topological_load_safetensor` method that is matching keys from the topology of the tree (and not only the exact adress of the weights) and give the ability to squeeze some nodes (like `DropoutAdapter` or `Chain`)

Even if I think it can solve most of the problems, there are still cases where such `topological_load` won't work.

One example would be to match together `Asymetric dropout` and `No dropout`

| Asymetric dropout | No dropout |
|---|---|
|![AsymDropout](https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20Linear1%5Blabel%3D%22Linear%22%5D%0A%20%20weight1%5Blabel%3D%22weight%22%5D%0A%20%20Linear2%5Blabel%3D%22Linear%22%5D%0A%20%20weight2%5Blabel%3D%22weight%22%5D%0A%20%20Linear1%20-%3E%20weight1%0A%20%20DropoutAdapter%20-%3E%20Linear2%0A%20%20Linear2%20-%3E%20weight2%0A%20%7D) | ![simple2linars](https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20Linear1%5Blabel%3D%22Linear_1%22%5D%0A%20%20weight1%5Blabel%3D%22weight%22%5D%0A%20%20Linear2%5Blabel%3D%22Linear_2%22%5D%0A%20%20weight2%5Blabel%3D%22weight%22%5D%0A%20%20Linear1%20-%3E%20weight1%0A%20%20Linear2%20-%3E%20weight2%0A%20%7D)|

This is because the keying of asymetric Dropout 
```
[
  "Linear.weight",
  "DropoutAdapter.Linear.weight",
]
```
will be the same no matter if it's `Linear_1` or `Linear_2` which has been dropped-out.
Rk: it seems that even if `state_dict.keys()` is following the orders of modules, saving it to safetensors will break this order and when we load it, we get alphabetical order.

So there is no way to make `topological_load_safetensor` works in such cases.

I'm not clear if those "asymetric dropout" situation happens often, or if there are similar situation where the same limitation occurs.

In practice, i feel most of the dropout are applied symmetrically to all the linear layers of a specific module, and this edge case is not a problem.

So does it worth it to work on a `topological_load_safetensor` approach ? considering that this function can raise an error if a conflict arise (like in the `Asymetric dropout` example) or is it needed to rework the keying ?

In order to structure my understanding, I've made a list of examples, criteria and keying strategies,  in https://docs.google.com/spreadsheets/d/1KIjUZQvQuKIZg4Ns-OaZLxRmtS-ZTWlu357Pj3drKq0/edit#gid=1934678375 I would be glad to get any feedback, more example, more criteria ...

**Reply by catwell** [reply]
@piercus I didn't look at the details yet but pinging @limiteinductive because he had a similar idea some time ago (with a specific algorithm in mind).

Note: we can't just reason about Dropout, because solving just the Dropout case is easy - we could special-case layers like Dropout so they are ignored in the keys, i.e. so that the state dict was generated as if they were ejected.

I'll make sure to look at this in more details soon, thank you!

**Reply by limiteinductive** [reply]
This is indeed something I've been contemplating for a while. A propos Dropout: It would make more sense to target activations instead of targeting linear layers, which would have the effect of not modifying the state_dict (PR incoming).

**Reply by limiteinductive** [reply]
My idea for the topological_state_dict is to use its similarity with knot theory and have an algorithm made of something that would be a composition of Reidermeister moves.

**Reply by limiteinductive** [reply]
What you propose is similar to saying that the topological_state_dict is invariant under the type I Reidermeister move
![image](https://github.com/finegrain-ai/refiners/assets/82658997/4f2a8ffc-2cad-4e4f-a9f2-cdad867d7ac8)

**Reply by catwell** [reply]
I change my mind on this everyday but: should we do the simple, useful but incomplete solution first?

(Digression: last time I changed my mind from "slightly against" to "slightly in favor" is watching [my friend Hisham's talk about Teal](https://fosdem.org/2024/schedule/event/fosdem-2024-2564-five-years-of-teal-minimalism-versus-growth-in-language-design/) last Sunday at FOSDEM, where he made a series of such choices.)

The solution being: have hacks in chain so we can have layers that either "masquerade" as other layers (i.e. for instance if you replace a layer A by B, still name it A in the dict) or are "transparent" (skipped) in the state dict. Those hacks are easy to implement and I already use them when prototyping stuff to avoid converting weights.

That would work for e.g. Dropout, but it would not work for everything though, and it would require manual effort.

**Comment by limiteinductive** [comment]
The big fundamental question is, should the keys of the state_dict be human-readable?

**Reply by catwell** [reply]
My POV on this is it helps a lot when it is. But I am biased by my experience writing and debugging code that converts state dicts between formats. :) Most users don't care.

**Reply by deltheil** [reply]
> [...] should the keys of the state_dict be human-readable?

Cherry on the cake (agreed on the benefit when debugging/inspecting)
